fpga_interchange/site_arch.cc

```diff
@@ -0,0 +1,415 @@
+
+SiteArch::SiteArch(const SiteInformation *site_info) : ctx(site_info->ctx), site_info(site_info)
+{
+    // Build list of input and output site ports
+    //
+    // FIXME: This doesn't need to be computed over and over, move to
+    // arch/chip db.
+    const TileTypeInfoPOD &tile_type = loc_info(&site_info->chip_info(), *site_info);
+    PipId pip;
+    pip.tile = site_info->tile;
+    for (size_t pip_index = 0; pip_index < tile_type.pip_data.size(); ++pip_index) {
+        if (tile_type.pip_data[pip_index].site != site_info->site) {
+            continue;
+        }
+
+        pip.index = pip_index;
+
+        if (!site_info->is_site_port(pip)) {
+            continue;
+        }
+
+        WireId src_wire = ctx->getPipSrcWire(pip);
+        if (site_info->is_wire_in_site(src_wire)) {
+            output_site_ports.push_back(pip);
+        } else {
+            input_site_ports.push_back(pip);
+        }
+    }
+
+    // Create list of out of site sources and sinks.
+
+    for (CellInfo *cell : site_info->cells_in_site) {
+        for (const auto &pin_pair : cell->cell_bel_pins) {
+            const PortInfo &port = cell->ports.at(pin_pair.first);
+            if (port.net != nullptr) {
+                nets.emplace(port.net, SiteNetInfo{port.net});
+            }
+        }
+    }
+
+    for (auto &net_pair : nets) {
+        NetInfo *net = net_pair.first;
+        SiteNetInfo &net_info = net_pair.second;
+
+        // All nets require drivers
+        NPNR_ASSERT(net->driver.cell != nullptr);
+
+        bool net_driven_out_of_site = false;
+        if (net->driver.cell->bel == BelId()) {
+            // The driver of this site hasn't been placed, so treat it as
+            // out of site.
+            out_of_site_sources.push_back(SiteWire::make(site_info, PORT_OUT, net));
+            net_info.driver = out_of_site_sources.back();
+            net_driven_out_of_site = true;
+        } else {
+            if (!site_info->is_bel_in_site(net->driver.cell->bel)) {
+
+                // The driver of this site has been placed, it is an out
+                // of site source.
+                out_of_site_sources.push_back(SiteWire::make(site_info, PORT_OUT, net));
+                // out_of_site_sources.back().wire = ctx->getNetinfoSourceWire(net);
+                net_info.driver = out_of_site_sources.back();
+
+                net_driven_out_of_site = true;
+            } else {
+                net_info.driver = SiteWire::make(site_info, ctx->getNetinfoSourceWire(net));
+            }
+        }
+
+        if (net_driven_out_of_site) {
+            // Because this net is driven from a source out of the site,
+            // no out of site sink is required.
+            continue;
+        }
+
+        // Examine net to determine if it has any users not in this site.
+        bool net_used_out_of_site = false;
+        for (const PortRef &user : net->users) {
+            if (user.cell == nullptr) {
+                // This is pretty weird!
```

Comment:
is this a case you are seeing? I don't think it's something that should ever be happening

Reply:
I'll change this to a NPNR_ASSERT then.  The problem with places where `Type *` appear, you can never be sure if a nullptr is allowed, or what.

Reply:
Fixed.